### PR TITLE
Watermark support

### PR DESCRIPTION
**Pull request by juliushaertl**
Based on #553 (needs rebase and base branch update after that one is merged)

- Implement watermarking support
- Created vue settings for new options, others can be migrated afterwards in a separate PR

### CODE REVIEW

File: lib/Controller/WopiController.php [hunk 1]
```diff
@@ -176,6 +182,20 @@ public function checkFileInfo($fileId, $access_token) {
 			$response['TemplateSaveAs'] = $file->getName();
 		}
 
+		if ($this->shouldWatermark($isPublic, $wopi->getEditorUid(), $fileId, $wopi)) {
+			$replacements = [
+				'userId' => $wopi->getEditorUid(),
+				'date' => (new \DateTime())->format('Y-m-d H:i:s'),
+				'themingName' => \OC::$server->getThemingDefaults()->getName(),
+				'themingName' => \OC::$server->getThemingDefaults()->getName(),
```
Comment (kesselb) [inline lib/Controller/WopiController.php]: Looks like a typo

Comment (juliushaertl) [reply lib/Controller/WopiController.php]: Ah true, thanks :+1:

**Comment by juliushaertl** [comment]
Settings are for now stored in the `files` app context in app config, so they can be shared with other apps that support watermarking:

- **watermark_enabled** (String) `'yes'|'no'` Watermarking is enabled
- **watermark_text** (String) Text that should be displayed in the watermark. Replacing placeholders needs to be handled in the implementing app.
- **watermark_shareAll**  (String) `'yes'|'no'` Watermark all internal shares
- **watermark_shareRead** (String) `'yes'|'no'` Watermark all internal read only shares
- **watermark_linkSecure** (String) `'yes'|'no'` Watermark all link shares with secure view enabled
- **watermark_linkAll** (String) `'yes'|'no'` Watermark all link shares 
- **watermark_linkRead** (String) `'yes'|'no'` Watermark all read only link shares 
- **watermark_linkTags** (String) `'yes'|'no'` Watermark all shares tagged with either one of the tags in `watermark_linkTagsList`
- **watermark_linkTagsList** (String) List of system tags to apply a watermark on link shares (Separated by `,`)
- **watermark_allGroups** (String) `'yes'|'no'` Enforce watermarking for users of a specific group, defined in `watermark_allGroupsList`
- **watermark_allGroupsList** (String) List of groups to apply a watermark (Separated by `,`)
- **watermark_allTags** (String) `'yes'|'no'` Watermark all files tagged with either one of the tags in `watermark_allTagsList`
- **watermark_allTagsList** List of system tags to apply a watermark on files (Separated by `,`)

**Comment by rullzer** [comment]
That is a lot of options :P

**Pull request merged**